<H1> Pandas DataFrame Part2 </H1>

In [2]:
import numpy as np
import pandas as pd

In [3]:
#Index level
outside=['G1','G1','G1','G2','G2','G2']
inside=[1,2,3,1,2,3]
hier_index=list(zip(outside,inside))
hier_index=pd.MultiIndex.from_tuples(hier_index)

In [4]:
list(zip(outside,inside))


[('G1', 1), ('G1', 2), ('G1', 3), ('G2', 1), ('G2', 2), ('G2', 3)]

<h3> "MultiIndex" takes tuples/arrays in zip format to any number of arrays; 
Syntax: list(zip(*arrays)) ; and it will work as rows for matrix;
MultiIndex takes three arguments : size of matrix,rows,columns</h3>

In [5]:
pd.MultiIndex.from_tuples(hier_index)

MultiIndex([('G1', 1),
            ('G1', 2),
            ('G1', 3),
            ('G2', 1),
            ('G2', 2),
            ('G2', 3)],
           )

In [6]:
pd.MultiIndex.from_tuples

<bound method MultiIndex.from_tuples of <class 'pandas.core.indexes.multi.MultiIndex'>>

In [7]:
df=pd.DataFrame(np.random.randn(6,2),hier_index,['A','B'])

In [8]:
df

A         B
G1 1  0.038786 -0.645830
   2 -0.093006  1.247385
   3  1.050611 -2.386732
G2 1 -1.437198 -0.372119
   2 -0.878282  0.278719
   3 -1.402327  0.551820

In [9]:
df.loc['G1']

A         B
1  0.038786 -0.645830
2 -0.093006  1.247385
3  1.050611 -2.386732

In [10]:
df.loc['G1'].loc[1]


A    0.038786
B   -0.645830
Name: 1, dtype: float64

In [11]:
df.index.names

FrozenList([None, None])

<h3> Here, we donot have any index names, thus we are assigning name to our rows like given below: </h3>

In [12]:
df.index.names=['Groups','Number']

In [13]:
df

A         B
Groups Number                    
G1     1       0.038786 -0.645830
       2      -0.093006  1.247385
       3       1.050611 -2.386732
G2     1      -1.437198 -0.372119
       2      -0.878282  0.278719
       3      -1.402327  0.551820

<H3> We want to grab "0.278719" of DataFrame which is on index Row G2 of row 2nd index and column B, let's checkout how we can do it. </H3>

In [15]:
df.loc['G2']

A         B
Number                    
1      -1.437198 -0.372119
2      -0.878282  0.278719
3      -1.402327  0.551820

In [16]:
df.loc['G2'].loc[2]

A   -0.878282
B    0.278719
Name: 2, dtype: float64

In [17]:
df.loc['G2'].loc[2]['B']

0.27871948798222634

<h3> "cross section" in Multilevel Index of Pandas as it has the ability to cross and go inside the DataFrame; Syntax "DataFrame.xs('Index_name')"
Like if we want to grab index with level 1 then we can do it by specifying index and level by using cross section</h3>

In [18]:
df.xs('G1')

A         B
Number                    
1       0.038786 -0.645830
2      -0.093006  1.247385
3       1.050611 -2.386732

In [20]:
df.xs(1,level='Number')

A         B
Groups                    
G1      0.038786 -0.645830
G2     -1.437198 -0.372119

In [24]:
df.xs(2,level='Number')

A         B
Groups                    
G1     -0.093006  1.247385
G2     -0.878282  0.278719

<h1> Missing Data in DataFrame </h1>

In [25]:
d={'A':[1,2,np.nan],'B':[5,np.nan,np.nan],'C':[1,2,3]}

In [26]:
d

{'A': [1, 2, nan], 'B': [5, nan, nan], 'C': [1, 2, 3]}

In [27]:
df1=pd.DataFrame(d)

In [28]:
df1

A    B  C
0  1.0  5.0  1
1  2.0  NaN  2
2  NaN  NaN  3

In [30]:
df1.dropna()   #it specifies that donot contain of null values and by default it takes axis=0 which means Rows of matrix

A    B  C
0  1.0  5.0  1

In [31]:
df1.dropna(axis=1)

C
0  1
1  2
2  3

In [32]:
df1.dropna(axis=0)

A    B  C
0  1.0  5.0  1

In [35]:
df1.dropna(thresh=2)   #drop non-NA values in which two values are there with threshold of 2

A    B  C
0  1.0  5.0  1
1  2.0  NaN  2

In [37]:
df1.fillna("fill")           #It will fill missing values by "Fill" but to change permanently, put "inplace=True"

A     B  C
0     1     5  1
1     2  fill  2
2  fill  fill  3

In [38]:
df1['A']

0    1.0
1    2.0
2    NaN
Name: A, dtype: float64

In [41]:
df1['A'].fillna(value=df['A'].mean(),inplace=True)

In [42]:
df1

A    B  C
0  1.000000  5.0  1
1  2.000000  NaN  2
2 -0.453569  NaN  3

<h3> Group By function in pandas . GroupBy allows you to group together rows based off a column and perform an aggregate fucntion on them </h3>


<h1> GroupBy - Pandas </h1>

In [43]:
data={'Company':['GooG','GooG','MSFT','MSFT','FB','FB'],
      'Person':['Sam','Charlie','Amy','Vaneesa','Carl','Sarah'],
     'Sales':[200,120,340,124,243,350]}

In [44]:
data

{'Company': ['GooG', 'GooG', 'MSFT', 'MSFT', 'FB', 'FB'],
 'Person': ['Sam', 'Charlie', 'Amy', 'Vaneesa', 'Carl', 'Sarah'],
 'Sales': [200, 120, 340, 124, 243, 350]}

In [45]:
df_groupBy=pd.DataFrame(data)

In [46]:
df_groupBy


Company   Person  Sales
0    GooG      Sam    200
1    GooG  Charlie    120
2    MSFT      Amy    340
3    MSFT  Vaneesa    124
4      FB     Carl    243
5      FB    Sarah    350

In [50]:
by_Company=df_groupBy.groupby('Company')

In [53]:
calc=by_Company.mean()  #Applied aggregate function on GroupBy

In [52]:
calc

Sales
Company       
FB       296.5
GooG     160.0
MSFT     232.0

In [55]:
by_Company.std()

Sales
Company            
FB        75.660426
GooG      56.568542
MSFT     152.735065

In [57]:
by_Company.sum().loc['FB']

Sales    593
Name: FB, dtype: int64

<h3> we can perform whole operation in one line like shown below. </h3>


In [59]:
df_groupBy.groupby('Company').mean().loc['GooG']

Sales    160.0
Name: GooG, dtype: float64

In [60]:
df_groupBy.groupby('Company').count()

Person  Sales
Company               
FB            2      2
GooG          2      2
MSFT          2      2

In [61]:
    df_groupBy.groupby('Company').max()   # alphabetical sorting

Person  Sales
Company                
FB         Sarah    350
GooG         Sam    200
MSFT     Vaneesa    340

In [63]:
df_groupBy.groupby('Company').describe()

Sales                                                        
        count   mean         std    min     25%    50%     75%    max
Company                                                              
FB        2.0  296.5   75.660426  243.0  269.75  296.5  323.25  350.0
GooG      2.0  160.0   56.568542  120.0  140.00  160.0  180.00  200.0
MSFT      2.0  232.0  152.735065  124.0  178.00  232.0  286.00  340.0

In [65]:
df_groupBy.groupby('Company').describe().transpose()

Company              FB        GooG        MSFT
Sales count    2.000000    2.000000    2.000000
      mean   296.500000  160.000000  232.000000
      std     75.660426   56.568542  152.735065
      min    243.000000  120.000000  124.000000
      25%    269.750000  140.000000  178.000000
      50%    296.500000  160.000000  232.000000
      75%    323.250000  180.000000  286.000000
      max    350.000000  200.000000  340.000000

In [66]:
df_groupBy.groupby('Company').describe().transpose()['GooG']

Sales  count      2.000000
       mean     160.000000
       std       56.568542
       min      120.000000
       25%      140.000000
       50%      160.000000
       75%      180.000000
       max      200.000000
Name: GooG, dtype: float64